### PR TITLE
luci-admin-status:Add kicking clients function

### PR DESCRIPTION
**Pull request by 805447391**
As the title, there is a more convenient way to kick some WIFI clients out.
image
Using `ubus call hostapd.wlan1 del_client "{'addr':'AC:0D:1B:D0:60:29', 'reason':5, 'deauth':true, 'ban_time':60000}"`
Deafult ban time: 60s
DeAuth: true
Thanks for reading.

### CODE REVIEW

File: modules/luci-base/luasrc/view/wifi_assoclist.htm [hunk 1]
```diff
@@ -68,13 +71,21 @@
 	);
 //]]></script>
 
+<script type="text/javascript">//<![CDATA[
+	function kick_client(iface, mac) {
```
Comment (jow-) [inline modules/luci-base/luasrc/view/wifi_assoclist.htm]: Please add the function to the existing `<script>` block, rename it to `function handleDisconnect(ev) {` and obtain iface and bssid using:

```
var iface = ev.target.getAttribute('data-iface'),
    bssid = ev.target.getAttribute('data-bssid');
```

File: modules/luci-base/luasrc/view/wifi_assoclist.htm [hunk 2]
```diff
@@ -68,13 +71,21 @@
...
+//]]></script>
+
 <div class="table" id="wifi_assoclist_table">
 	<div class="tr table-titles">
 		<div class="th nowrap"><%:Network%></div>
 		<div class="th hide-xs"><%:MAC-Address%></div>
 		<div class="th nowrap"><%:Host%></div>
 		<div class="th nowrap"><%:Signal%> / <%:Noise%></div>
 		<div class="th nowrap"><%:RX Rate%> / <%:TX Rate%></div>
+		<div class="th nowrap">Kick</div>
```
Comment (jow-) [inline modules/luci-base/luasrc/view/wifi_assoclist.htm]: Please change this to `<%:Disconnect%>`

File: modules/luci-mod-status/luasrc/controller/admin/status.lua [hunk 3]
```diff
@@ -6,6 +6,7 @@ module("luci.controller.admin.status", package.seeall)
 
 function index()
 	entry({"admin", "status", "overview"}, template("admin_status/index"), _("Overview"), 1)
+	entry({"admin", "status", "overview", "kick"}, post("act_kick")).leaf = true
```
Comment (jow-) [inline modules/luci-mod-status/luasrc/controller/admin/status.lua]: Please rename the function to `action_disconnect` and the url part to `disconnect`.

File: modules/luci-base/luasrc/view/wifi_assoclist.htm [hunk 4]
```diff
@@ -58,7 +58,10 @@
 							E('span', wifirate(bss, true)),
 							E('br'),
 							E('span', wifirate(bss, false))
-						])
+						]),
+						'<input class="cbi-button cbi-input-remove" type="button" value="Kick" onclick="kick_client(\'%s\', \'%s\')" />'.format(
+							bss.ifname, 
+							bss.bssid)
 					]);
```
Comment (jow-) [inline modules/luci-base/luasrc/view/wifi_assoclist.htm]: Please change this to:

```
E('input', {
    type: 'button',
    class: 'cbi-button cbi-button-remove',
    value: '<%:Disconnect%>',
    'data-bssid': bss.bssid,
    'data-ifname': bss.ifname,
    click: handleDisconnect
})
```

File: modules/luci-base/luasrc/view/wifi_assoclist.htm [hunk 5]
```diff
@@ -68,13 +71,21 @@
 	);
 //]]></script>
 
+<script type="text/javascript">//<![CDATA[
+	function kick_client(iface, mac) {
+		(new XHR()).post('<%=url('admin/status/overview/kick')%>/', {token: '<%=token%>', iface: iface, mac: mac}, function(x) {
```
Comment (jow-) [inline modules/luci-base/luasrc/view/wifi_assoclist.htm]: Please change `kick` to `disconnect` and `mac` to `bssid`.

File: modules/luci-mod-status/luasrc/controller/admin/status.lua [hunk 6]
```diff
@@ -183,3 +184,13 @@ function action_nameinfo(...)
 		limit = 1000
 	}) or { })
 end
+
+function act_kick()
+	local iface = luci.http.formvalue("iface")
+	local mac = luci.http.formvalue("mac")
```
Comment (jow-) [inline modules/luci-mod-status/luasrc/controller/admin/status.lua]: Please change `mac` to `bssid`.

File: modules/luci-mod-status/luasrc/controller/admin/status.lua [hunk 7]
```diff
@@ -183,3 +184,13 @@ function action_nameinfo(...)
...
+	if iface and mac then
+		local cmd = string.format("ubus call hostapd.%s del_client \"{'addr':'%s', 'reason':5, 'deauth':true, 'ban_time':60000}\"", iface, mac)
```
Comment (jow-) [inline modules/luci-mod-status/luasrc/controller/admin/status.lua]: This is a textbook example of a root command injection vulnerability. Please escape `iface` and `bssid` through [`luci.util.shellquote()`](http://htmlpreview.github.io/?http://raw.githubusercontent.com/openwrt/luci/master/documentation/api/modules/luci.util.html#shellquote)

**Comment by jow-** [comment]
Hi, thanks for you contribution! A couple of comments below.

File: modules/luci-mod-status/luasrc/controller/admin/status.lua (hunk 7 above)
Comment (jow-) [reply modules/luci-mod-status/luasrc/controller/admin/status.lua]: Actually, there's an even better way. LuCI has a native ubus bindings, so you can replace the entire shell code with this:

```
luci.util.ubus("hostapd.%s" % iface, "del_client", {
  addr = bssid,
  deauth = true,
  reason = 5,
  ban_time = 60000
})
```

File: modules/luci-base/luasrc/view/wifi_assoclist.htm (hunk 4 above)
Comment (805447391) [reply modules/luci-base/luasrc/view/wifi_assoclist.htm]: image
image

Why not send a HTTP request when click 'disconnect' button using you way ? excuse me.

Comment (805447391) [reply modules/luci-base/luasrc/view/wifi_assoclist.htm]: here is two pictures.

Comment (jow-) [reply modules/luci-base/luasrc/view/wifi_assoclist.htm]: I guess you're not using the latest master. You need commit https://github.com/openwrt/luci/commit/f4c39dd6ff6f60dc48982d822faa6d9a85c2e4d7 fir this to work correctly.

**Comment by 805447391** [comment]
changed.

**Comment by 805447391** [comment]
Why is there still a review? pls.

**Comment by jow-** [comment]
Merged with various fixes and changes in 09ac96ede. Thanks!

**Pull request closed without merging**